ipa-metrics/src/store.rs

```diff
@@ -0,0 +1,224 @@
+use std::{borrow::Borrow, hash::BuildHasher};
+
+use hashbrown::hash_map::RawEntryMut;
+use rustc_hash::FxBuildHasher;
+
+use crate::{key::OwnedMetricName, kind::CounterValue, MetricName};
+
+/// A basic store. Currently only supports counters.
+/// Counters and other metrics are stored to optimize writes. That means, one lookup
+/// per write. The cost of assembling the total count across all dimensions is absorbed
+/// by readers
+#[derive(Clone, Debug)]
+pub struct Store {
+    counters: hashbrown::HashMap<OwnedMetricName, CounterValue, FxBuildHasher>,
+}
+
+impl Default for Store {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Store {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            counters: hashbrown::HashMap::with_hasher(FxBuildHasher),
+        }
+    }
+
+    pub fn merge(&mut self, other: Self) {
+        for (k, v) in other.counters {
+            let hash_builder = self.counters.hasher();
+            let hash = hash_builder.hash_one(&k);
+            *self
+                .counters
+                .raw_entry_mut()
+                .from_hash(hash, |other| other.eq(&k))
+                .or_insert(k, 0)
+                .1 += v;
+        }
+    }
+
+    pub fn counter<'a, const LABELS: usize, B: Borrow<MetricName<'a, LABELS>>>(
+        &'a mut self,
+        key: B,
+    ) -> CounterHandle<'a, LABELS> {
+        let key = key.borrow();
+        let hash_builder = self.counters.hasher();
+        let hash = hash_builder.hash_one(key);
+        let entry = self
+            .counters
+            .raw_entry_mut()
+            .from_hash(hash, |key_found| key_found.eq(key));
+        match entry {
+            RawEntryMut::Occupied(slot) => CounterHandle {
+                val: slot.into_mut(),
+            },
+            RawEntryMut::Vacant(slot) => {
+                let (_, val) = slot.insert_hashed_nocheck(hash, key.to_owned(), Default::default());
+                CounterHandle { val }
+            }
+        }
+    }
+
+    /// Returns the value for the specified metric taking into account
+    /// its dimensionality. That is (foo, dim1 = 1, dim2 = 2) will be
+    /// different from (foo, dim1 = 1).
```

Comment:
This description is confusing to me. Isn't the behavior of the partial match logic that a query for `(foo, dim1 = 1)` will include data for metric names `(foo, dim1 = 1)`, `(foo, dim1 = 1, dim2 = 2)`.

Taking a stab at rewording it:
> Returns the value for the specified metric, limited by any specified dimensions, but not by any unspecified dimensions. If metric `foo` has dimensions `dim1` and `dim2`, a query for `(foo, dim1 = 1)` will sum the counter values having `dim1 = 1` and any value of `dim2`.